### PR TITLE
Create MongoDB Resource Manager

### PR DESCRIPTION
**Pull request by Polber**
This PR introduces several changes...
- There is now a MongoDB Resource Manager for writing integration tests.
- There is an integration test for the MongoDBToBigQuery template
- The TestContainers framework has been added for writing resource managers that are not Google Cloud products (i.e. MongoDB)

Signed-off-by: Jeffrey Kinard <jeff@thekinards.com>

Fixes #507

### CODE REVIEW

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * Utilities for {@link com.google.cloud.teleport.it.mongodb.MongoDBResourceManager}
+ * implementations.
+ */
+final class MongoDBResourceManagerUtils {
+
+  private static final int MAX_DATABASE_NAME_LENGTH = 63;
+  private static final Pattern ILLEGAL_DATABASE_NAME_CHARS =
+      Pattern.compile("[\\/\\\\. \"\0$]"); // i.e. [/\. "$]
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java:31]: nit Do we have a source/link to cite the illegal chars? Good for posterity

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  private static final String REPLACE_DATABASE_NAME_CHAR = "-";
+  private static final int MIN_COLLECTION_NAME_LENGTH = 1;
+  private static final int MAX_COLLECTION_NAME_LENGTH = 99;
+  private static final Pattern ILLEGAL_COLLECTION_CHARS = Pattern.compile("[$\0]");
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java:35]: idem

File: it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.GenericContainer;
+
+/** Interface for managing TestContainers resources in integration tests. */
+public abstract class TestContainerResourceManager<T extends GenericContainer<?>> {
+  private static final Logger LOG = LoggerFactory.getLogger(TestContainerResourceManager.class);
+
+  private final T container;
+  private final boolean usingStaticContainer;
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java:27]: Giving some context on what is static container will he helpful. Maybe at class level?

File: it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        LOG.error("Failed to close TestContainer resources. {}", e.getMessage());
+        return false;
+      }
+    } else {
+      LOG.info("This manager was configured to use a static resource that will not be cleaned up.");
+    }
+
+    LOG.info("TestContainers manager successfully cleaned up.");
+    return true;
+  }
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java:100]: nit / might be 100% preference based but the code would look cleaner short circuiting on `usingStaticContainer` e.g.,

```suggestion
  protected boolean cleanupAll() {
    if (usingStaticContainer) {
       LOG.info("This manager was configured to use a static resource that will not be cleaned up.");
       return true;
    }
    
    LOG.info("Attempting to cleanup TestContainers manager.");
    try {
      container.close();
      LOG.info("TestContainers manager successfully cleaned up.");
      return true;
    } catch (Exception e) {
      LOG.error("Failed to close TestContainer resources. {}", e.getMessage());
      return false;
    }
  }
```

File: it/src/test/java/com/google/cloud/teleport/it/common/ResourceManagerUtilsTest.java [hunk 5]
```diff
@@ -80,7 +80,7 @@ public void testGenerateResourceIdShouldReplaceUnderscoreWithHyphen() {
   }
 
   @Test
-  public void testGenerateResourceIdShouldReplaceUpperCaseLettersWithLowerCase() {
+  public void testGenerateResourceIdShouldReplaceUpperCaseLettersWithHyphen() {
```
Comment (bvolpato) [inline it/src/test/java/com/google/cloud/teleport/it/common/ResourceManagerUtilsTest.java:83]: Any particular reason? I think lowering first seems much better

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+  private static final int MONGODB_INTERNAL_PORT = 27017;
+
+  private final MongoClient mongoClient;
+  private final String databaseName;
+  private final String connectionString;
+  private final boolean usingStaticDatabase;
+
+  private DefaultMongoDBResourceManager(DefaultMongoDBResourceManager.Builder builder) {
+    this(
+        null,
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:65]: nit but this is recommended
```suggestion
        /*mongoClient=*/ null,
```

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    return true;
+  }
+
+  @Override
+  public synchronized FindIterable<Document> readCollection(String collectionName) {
+    LOG.info("Reading all documents from {}.{}", databaseName, collectionName);
+
+    FindIterable<Document> documents;
+    try {
+      documents = getMongoDBCollection(collectionName, false).find();
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:201]:
```suggestion
      documents = getMongoDBCollection(collectionName, /*createCollection=*/ false).find();
```

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+  @Override
+  public synchronized boolean insertDocuments(String collectionName, List<Document> documents) {
+    LOG.info(
+        "Attempting to write {} documents to {}.{}.",
+        documents.size(),
+        databaseName,
+        collectionName);
+
+    try {
+      getMongoDBCollection(collectionName, true).insertMany(documents);
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:184]:
```suggestion
      getMongoDBCollection(collectionName, /*createCollection=*/ true).insertMany(documents);
```

File: v2/mongodb-to-googlecloud/pom.xml [hunk 9]
```diff
@@ -29,13 +29,16 @@
 
     <artifactId>mongodb-to-googlecloud</artifactId>
 
+    <properties>
+    <google-truth.version>1.0.1</google-truth.version>
+    </properties>
+
```
Comment (bvolpato) [inline v2/mongodb-to-googlecloud/pom.xml:35]:
```suggestion
```

File: v2/mongodb-to-googlecloud/pom.xml [hunk 10]
```diff
@@ -45,6 +48,14 @@
             <artifactId>common</artifactId>
             <version>${project.version}</version>
         </dependency>
+
+        <!-- Test Dependencies -->
+        <dependency>
+            <groupId>com.google.truth</groupId>
+            <artifactId>truth</artifactId>
+            <version>${google-truth.version}</version>
```
Comment (bvolpato) [inline v2/mongodb-to-googlecloud/pom.xml:56]:
```suggestion
            <version>${truth.version}</version>
```

File: v2/pom.xml [hunk 11]
```diff
@@ -495,6 +495,7 @@
                 <artifactId>maven-surefire-plugin</artifactId>
                 <version>${surefire.version}</version>
                 <configuration>
+                    <forkCount>0</forkCount>
```
Comment (bvolpato) [inline v2/pom.xml:498]: nit wrong ident
```suggestion
                  <forkCount>0</forkCount>
```

File: v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+  }
+
+  @After
+  public void tearDownClass() {
+    boolean producedError = false;
+
+    try {
+      mongoDbClient.cleanupAll();
+    } catch (Exception e) {
+      LOG.error("Failed to delete MongoDB resources. {}", e.getMessage());
```
Comment (bvolpato) [inline v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java:130]:
```suggestion
      LOG.error("Failed to delete MongoDB resources.", e);
```

File: v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+      producedError = true;
+    }
+
+    try {
+      bigQueryClient.cleanupAll();
+    } catch (Exception e) {
+      LOG.error("Failed to delete MongoDB resources. {}", e.getMessage());
```
Comment (bvolpato) [inline v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java:137]:
```suggestion
      LOG.error("Failed to delete MongoDB resources.", e);
```

**Review by bvolpato — CHANGES_REQUESTED** [review]
Looking pretty good! Good stuff :)

Dropped a few comments. Can you PTAL?

File: it/src/main/java/com/google/cloud/teleport/it/bigquery/DefaultBigQueryResourceManager.java [hunk 14]
```diff
@@ -255,32 +250,30 @@ public synchronized void write(String tableName, List<RowToInsert> rows) {
...
     // Read all the rows from the table given by tableId
+    TableResult results;
     try {
-      TableResult results = bigQuery.listTableData(table.getTableId());
-      for (FieldValueList row : results.getValues()) {
-        immutableListBuilder.add(row);
-      }
+      String query = "SELECT TO_JSON_STRING(t) FROM `";
+      query += String.join(".", projectId, datasetId, tableName);
+      query += "` AS t;";
```
[the remark this answers is not in the document]
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/bigquery/DefaultBigQueryResourceManager.java:263]: Changed to single line string.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+  public synchronized boolean createCollection(String collectionName) {
+    LOG.info("Creating collection using tableName '{}'.", collectionName);
+
+    try {
+      // Check to see if the Collection exists
+      if (collectionExists(collectionName)) {
+        return false;
+      }
+      // The Collection does not exist in the database, create it and return true.
+      getDatabase().getCollection(collectionName);
```
[the remark this answers is not in the document]
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:134]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    LOG.info(
+        "Attempting to write {} documents to {}.{}.",
+        documents.size(),
+        databaseName,
+        collectionName);
+
+    try {
+      getMongoDBCollection(collectionName, true).insertMany(documents);
+    } catch (Exception e) {
+      throw new MongoDBResourceManagerException("Error inserting document.", e);
```
[the remark this answers is not in the document]
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:186]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+
+    boolean producedError = false;
+
+    // First, delete the database if it was not given as a static argument
+    try {
+      if (!usingStaticDatabase) {
+        mongoClient.getDatabase(databaseName).drop();
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to delete MongoDB database {}. {}", databaseName, e.getMessage());
```
[the remark this answers is not in the document]
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:223]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+      producedError = true;
+    }
+
+    // Next, try to close the MongoDB client connection
+    try {
+      mongoClient.close();
+    } catch (Exception e) {
+      LOG.error("Failed to delete MongoDB client. {}", e.getMessage());
```
[the remark this answers is not in the document]
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:231]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java (hunk 4 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java:100]: I agree.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java (hunk 6 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:65]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java (hunk 7 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:201]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java (hunk 8 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java:184]: ack.

File: v2/mongodb-to-googlecloud/pom.xml (hunk 9 above)
Comment (Polber) [reply v2/mongodb-to-googlecloud/pom.xml:35]: Should be using v2/pom.xml for truth version anyway. Removed.

File: v2/mongodb-to-googlecloud/pom.xml (hunk 10 above)
Comment (Polber) [reply v2/mongodb-to-googlecloud/pom.xml:56]: ack.

File: v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java (hunk 12 above)
Comment (Polber) [reply v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java:130]: ack.

File: v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java (hunk 13 above)
Comment (Polber) [reply v2/mongodb-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/mongodb/templates/MongoDbToBigQueryIT.java:137]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+ *
+ * <p>The class is thread-safe.
+ */
+public class DefaultMongoDBResourceManager extends TestContainerResourceManager<GenericContainer<?>>
+    implements MongoDBResourceManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultBigtableResourceManager.class);
+  private static final String DEFAULT_MONGODB_CONTAINER_NAME = "mongo";
+  private static final String DEFAULT_MONGODB_CONTAINER_TAG = "4.0.10";
+  private static final int MONGODB_INTERNAL_PORT = 27017;
```
[the remark this answers is not in the document]
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/DefaultMongoDBResourceManager.java]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java (hunk 1 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java:31]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java (hunk 2 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManagerUtils.java:35]: ack.

File: it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java (hunk 3 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/testcontainers/TestContainerResourceManager.java:27]: ack.

File: it/src/test/java/com/google/cloud/teleport/it/common/ResourceManagerUtilsTest.java (hunk 5 above)
Comment (Polber) [reply it/src/test/java/com/google/cloud/teleport/it/common/ResourceManagerUtilsTest.java:83]: ack.

File: v2/pom.xml (hunk 11 above)
Comment (Polber) [reply v2/pom.xml]: forkCount has been removed.

File: it/pom.xml [hunk 20]
```diff
@@ -32,8 +32,14 @@
...
     <mockito.version>3.7.7</mockito.version>
     <mockito-inline.version>4.5.1</mockito-inline.version>
+    <mongodb.version>3.12.10</mongodb.version>
     <puppycrawl.version>8.7</puppycrawl.version>
     <dataflow-api.version>v1b3-rev20220920-2.0.0</dataflow-api.version>
+    <re2j.version>1.5</re2j.version>
+    <slf4j.version>1.7.25</slf4j.version>
+    <surefire.version>2.21.0</surefire.version>
+    <testcontainers.version>1.17.5</testcontainers.version>
+    <truth.version>1.0.1</truth.version>
```
Comment (bvolpato) [inline it/pom.xml:42]: If things are declared on the main pom.xml file, there's no reason to configure them here too. The properties from the main pom are inherited from the <parent> relationship.

**Review by bvolpato — APPROVED** [review]
LGTM. I've just left a suggestion so we can reduce the `it/pom.xml` delta.

File: it/pom.xml (hunk 20 above)
Comment (Polber) [reply it/pom.xml]: removed redundant versions

**Review by bvolpato — APPROVED** [review]

**Pull request merged**